### DISCUSSION TITLE
Allow `on` to be a `predicate: typing.Callable[[BaseException], bool]`

### DISCUSSION THREAD

**Discussion by dhirschfeld** [hynek/stamina#63] [Ideas]
I'm dealing with a library where I want to retry an exception only if the message contains a certain string.

I think the most flexible way to support this whilst being backward compatible would be to change the `on` argument to allow passing in a `predicate` callable - e.g.:
```
 on: type[Exception] | tuple[type[Exception], ...] | Callable[[BaseException], bool],
```
In the `retry` code you can then create the `predicate` like:
```python
if isinstance(on, Callable):
    predicate = on
else:
    predicate = lambda exc: isinstance(exc, on)
```
...and then in [`_RetryContextIterator`](https://github.com/hynek/stamina/blob/main/src/stamina/_core.py#L392) pass it through to `tenacity` as:
```python
        return cls(
            _name=name,
            _args=args,
            _kw=kw,
            _t_kw={
                "retry": _t.retry_if_exception(predicate),
<snip>
```

***xref:*** https://github.com/jd/tenacity/blob/main/tenacity/retry.py#L62-L78

**Comment by estahn** [comment]
I think this is reasonable. I would like to retry in case the downstream service started throttling (429's). I understand this is possible with tenacity, but not with stamina due to the above?

**Comment by hynek** [comment]
I would be open to a PR, but bear in mind I'm leaving for a vacation tomorrow so merging might take some time!

**Comment by dhirschfeld** [comment]
Enjoy the vacation! 🌴 

No rush on this (for me at least). I'll try to find the time to put in a PR. It should be fairly straightforward.... *I think!*

**Reply by hynek** [reply]
famous last words? ;)

**Comment by hynek** [comment]
This has been merged in #70 – thanks to @gsakkis!
